lectures/mle.md

```diff
@@ -0,0 +1,249 @@
+
+We further use [jax.vmap](https://jax.readthedocs.io/en/latest/_autosummary/jax.vmap.html) which vectorizes the given function i.e. the function acting upon scalar inputs can now be used with vector inputs.
+
+```{code-cell} ipython3
+dlogL = jax.vmap(jax.grad(logL))
+```
+
+```{code-cell} ipython3
+β = jnp.linspace(1, 20)
+
+fig, (ax1, ax2) = plt.subplots(2, sharex=True, figsize=(12, 8))
+
+ax1.plot(β, logL(β), lw=2)
+ax2.plot(β, dlogL(β), lw=2)
+
+ax1.set_ylabel(r'$log \mathcal{L(\beta)}$',
+               rotation=0,
+               labelpad=35,
+               fontsize=15)
+ax2.set_ylabel(r'$\frac{dlog \mathcal{L(\beta)}}{d \beta}$ ',
+               rotation=0,
+               labelpad=35,
+               fontsize=19)
+
+ax2.set_xlabel(r'$\beta$', fontsize=15)
+ax1.grid(), ax2.grid()
+plt.axhline(c='black')
+plt.show()
+```
+
+
+The plot shows that the maximum likelihood value (the top plot) occurs
+when $\frac{d \log \mathcal{L(\boldsymbol{\beta})}}{d \boldsymbol{\beta}} = 0$ (the bottom
+plot).
+
+Therefore, the likelihood is maximized when $\beta = 10$.
+
+We can also ensure that this value is a *maximum* (as opposed to a
+minimum) by checking that the second derivative (slope of the bottom
+plot) is negative.
+
+The Newton-Raphson algorithm finds a point where the first derivative is
+0.
+
+To use the algorithm, we take an initial guess at the maximum value,
+$\beta_0$ (the OLS parameter estimates might be a reasonable
+guess), then
+
+
+Please refer to [this section](https://python.quantecon.org/mle.html#mle-with-numerical-methods) for the detailed algorithm.
+
+Let's have a go at implementing the Newton-Raphson algorithm.
+
+First, we'll create a `PoissonRegressionModel`.
+
+```{code-cell} ipython3
+PoissonRegressionModel = namedtuple('PoissonRegressionModel', ['X', 'y'])
+
+def create_poisson_model(X, y):
+    n, k = X.shape
+    # Reshape y as a n_by_1 column vector
+    y = y.reshape(n, 1)
+    X, y = jax.device_put((X, y))
+    return PoissonRegressionModel(X=X, y=y)
+```
+
+
+The following function computes the `factorial`.
+
+```{code-cell} ipython3
+@jax.jit
+def _factorial(n):
+    return jax.lax.exp(jax.lax.lgamma(n + 1.0)).astype(int)
+
+jax_factorial = jax.vmap(_factorial)
+```
+
+
+Let's define the Poisson Regression's log likelihood function.
+
+```{code-cell} ipython3
+@jax.jit
+def poisson_logL(β, model):
+    y = model.y
+    μ = jnp.exp(jnp.dot(model.X, β))
+    return jnp.sum(model.y * jnp.log(μ) - μ - jnp.log(jax_factorial(y)))
+```
+
+
+To find the gradient of the `poisson_logL`, we again use [jax.grad](https://jax.readthedocs.io/en/latest/_autosummary/jax.grad.html).
+
+To find the Hessian, we can directly use [jax.jacfwd](https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html#jacobians-and-hessians-using-jacfwd-and-jacrev).
+
+```{code-cell} ipython3
+G_poisson_logL = jax.grad(poisson_logL)
+H_poisson_logL = jax.jacfwd(G_poisson_logL)
+```
```

Comment:
Perhaps add one line saying that the Hessian is the Jacobian of the gradient?

Why do we use `jacfwd` in the second derivative?  Why note `jax.jacobian`?

Reply:
> Why do we use jacfwd in the second derivative?

I referred that using this section of the docs: https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html#jacobians-and-hessians-using-jacfwd-and-jacrev. 

Moreover, from this issue: https://github.com/google/jax/issues/2344, it says: "Note that `jacobian` is currently just an alias for `jaxrev`."
.

Reply:
Thanks @Smit-create .  Actually this discussion is very helpful and I suggest we add it and then connect it to our choice to use `jaxfwd` in this particular case.  We can write:

According to [the documentation](https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html#jacobians-and-hessians-using-jacfwd-and-jacrev),

* `jacfwd` uses forward-mode automatic differentiation, which is more efficient for “tall” Jacobian matrices, while 
* `jacrev` uses reverse-mode, which is more efficient for “wide” Jacobian matrices. 
 
(The documentation also states that when matrices that are near-square, `jacfwd` probably has an edge over `jacrev`.)

Reply:
Thanks, added that.